crates/exex/exex/src/dyexex.rs

```diff
@@ -0,0 +1,32 @@
+//! Dynamically loading type-safe abstractions
+
+/// Dynamically loaded ExEx entrypoint, that accepts the [`ExExContext`](`reth_exex::ExExContext`)
+/// and returns a Future that will be polled by the [`ExExManager`](`reth_exex::ExExManager`).
+///
+/// ## Example usage:
+/// ```rust
+/// use std::future::Future;
+/// use reth_exex::{define_exex, ExExContext};
+/// use reth_node_api::FullNodeComponents;
+///
+/// // Create a function to produce ExEx logic
+/// async fn exex<Node: FullNodeComponents>(_ctx: ExExContext<Node>) -> eyre::Result<()> {
+///     Ok(())
+/// }
+///
+/// // Use the macro to generate the entrypoint function
+/// define_exex!(exex,<Node>);
```

Comment:
this example will not work, because `exex` returns `impl std::future::Future<Output = eyre::Result<()>>`, we need to return a future instead

can we also add a test for this functionality?

Reply:
I tested it through doc-test and by creating standalone dylib

Reply:
sorry, I should have clarified this: we want the same interface as currently `install_exex` has, so that users can do some setup in the async closure that they pass, for example like here https://github.com/paradigmxyz/reth-exex-examples/blob/a0077f1f77e9ec5984442747ddd5cec97bf1068c/backfill/src/main.rs#L285-L313

that is why the function that's passed to the macro should return a `Future` that resolves into another `Future`

Reply:
@shekhirin, sorry, missed that comment
I think that will be good

```rust
#[macro_export]
macro_rules! define_exex {
    ($user_fn:ident) => {
        #[no_mangle]
        pub extern "C" fn _launch_exex<Node: FullNodeComponents>(
            ctx: $crate::ExExContext<Node>,
        ) -> impl std::future::Future<
            Output = eyre::Result<impl Future<Output = eyre::Result<()>> + Send>,
        > {
            $user_fn(ctx)
        }
    };
}
```

Reply:
then you can do

```rust
//!  user's src/lib

use std::future::Future;

use reth_exex::{define_exex, ExExContext};
use reth_node_api::FullNodeComponents;

async fn exex_init<Node: FullNodeComponents>(
    ctx: ExExContext<Node>,
) -> eyre::Result<impl std::future::Future<Output = eyre::Result<()>>> {
    Ok(exex(ctx))
}

async fn exex<Node: FullNodeComponents>(ctx: ExExContext<Node>) -> eyre::Result<()> {
	Ok(())
}

// Use the macro to generate the unmangled _launch_exex function
define_exex!(exex_init);
```

Reply:
yep this is great. Maybe we can even make it work with closures too, so that you can do
```rust
//!  user's src/lib

use std::future::Future;

use reth_exex::{define_exex, ExExContext};
use reth_node_api::FullNodeComponents;

async fn exex<Node: FullNodeComponents>(ctx: ExExContext<Node>) -> eyre::Result<()> {
    Ok(())
}

// Use the macro to generate the unmangled _launch_exex function
define_exex!(async move |ctx| {
    Ok(exex(ctx))
});
```

But it can be done separately in another PR

Reply:
@shekhirin modified
(Not for closure, just basically to accept ExEx-style double-wrapped fut)